src/main.rs

```diff
@@ -47,55 +49,90 @@ fn compile(txt: &str) -> (String, Vec<usize>) {
     (out, map)
 }
 
-fn interp(prg: &[u8], map: Vec<usize>) {
-    let mut pc = 0;
-    let mut ptr = 0;
-    let mut cells = Vec::with_capacity(30000);
-    cells.resize(30000, 0u8);
-    let mut stdin = std::io::stdin().bytes();
-    let mut stdout = std::io::stdout();
-    while pc < prg.len() {
-        match prg[pc] as char {
-            '>' => {
-                if ptr == cells.len() {
-                    cells.push(0);
-                }
-                ptr += 1;
-            }
-            '<' => {
-                if ptr > 0 {
-                    ptr -= 1;
-                }
-            }
-            '+' => {
-                cells[ptr] = cells[ptr].wrapping_add(1);
-            }
-            '-' => {
-                cells[ptr] = cells[ptr].wrapping_sub(1);
+#[derive(Debug)]
+struct InterpCx {
+    pc: usize,
+    ptr: usize,
+    cells: Vec<u8>,
+    prog: Vec<u8>,
+    map: Vec<usize>,
+    stdin: Bytes<Stdin>,
+    stdout: Stdout,
+}
+
+fn interp(prog: &[u8], map: Vec<usize>) {
+    let mut icx = InterpCx {
+        pc: 0,
+        ptr: 0,
+        cells: Vec::from([0; 30000]),
+        map,
+        prog: prog.to_vec(),
+        stdin: std::io::stdin().bytes(),
+        stdout: std::io::stdout(),
+    };
+
+    // FIXME only use locations for loop heads.
+    #[cfg(tracermode = "hw")]
```

Comment:
@ltratt this is the part I mentioned in MM. `locs` and `mtt` need to be entirely absent if the JIT is not enabled.

`{}` doesn't allow this.

We could use and `Option` but what would the inner type be when the JIT is disabled?

Reply:
What does "JIT not enabled mean"? Do you mean "not when we're tracing" or ...?

Reply:
It means, we built ykbf without yorick support. I figured we'd like to keep that option open for performance comparisons.

Reply:
I don't think we care about building without Yorick support. For benchmarking that sort of thing, we'll probably have to fork and disable things manually anyway, as we're bound to need to slightly adjust the interpreter to help the trace optimiser as things go forward.

Reply:
OK. That means all of this `cfg` dance can go.

Instead of merging this into master, shall we start a `yorick` branch and keep `master` "vanilla" (a baseline).

Reply:
It's `ykbf` for a reason :) Master should be Yk. We can always retrospectively create a "nonyk" branch off the current master commit later if we want to.

Reply:
OK.

So please continue your review, ignoring the `#[cfg(tracermode)]` stuff.

Reply:
e474194 removes the ability to build without the JIT.